### DISCUSSION TITLE
Why having setup key set implies opt = true?

### DISCUSSION THREAD

**Discussion by oblitum** [wbthomason/packer.nvim#441] [Q&A] [answered]
Hi. I have all my plugins correctly lazy loading by precisely making correct usage of `setup`, `config`, `opt`, `requires`, `after` and the undocumented `wants`. Everything is working as expected except for one thing. I forgot to replace some `config`s for `setup`s because some plugins require some configuration variables to be set before they're loaded. So I did it, but this at the same time implies they all change to `opt = true`. I don't understand this change in defaults due to setting a `setup` function, what's the rationale for this?? Plugins are `opt = false` by default, now if they have a `setup`, they become `opt = true` by default, it doesn't make sense to me.

**Comment by oblitum** [comment]
~~Most of the plugins that got moved to `opt/` are filetype plugins. Despite they now having turned into opt plugins, _they are_ correctly loading when opening a file of the given filetype, I got a bit confused here as I assumed opt plugins to load only when setting some trigger key, like `keys`, `cmd`, `ft`, etc. As they changed to opt, but I didn't set any key but `setup`, dunno why they are correctly lazy loading at all.~~

Nevermind. This is even weirder... despite they being moved to `opt/` due to setting `setup`, they are still being sourced!

So they got moved to `opt/` but they're not opt because they're still being sourced.

**Comment by nanotee** [comment, marked as answer]
My guess would be that since `packer` can't control when `start` packages are loaded, `setup` makes a package `opt` to ensure that your settings are loaded *before* said package

**Reply by oblitum** [reply]
That makes total sense now nanotee.... so what I just described above makes sense. The plugins both become opt but are loaded regardless. It's kinda grey area how to label them.

**Reply by wbthomason** [reply]
As `nanotee` notes, the reason for this is because we have to ensure that the `setup` key runs before the plugin is loaded, but native packages only allows this control for `opt` plugins. So, we make all plugins with a `setup` key `opt`, but then - if they have **no other reason to be `opt`** (e.g. no lazy-loading) - we `packadd` them immediately after their `setup` keys to ensure that they're eagerly loaded. Setting something like `opt = true` implies that the user has an explicit need for the plugin to **not** be eagerly loaded, and so we don't auto load them in that case. 

It is worth noting that an `opt` plugin is not necessarily a lazy-loaded plugin (as in this case).

I hope this helps, and sorry for the confusion!

**Reply by oblitum** [reply]
Thanks for the official clarification. This is exactly as we ended up deducing.

**Reply by wbthomason** [reply]
Yep - @nanotee had it covered, I just wanted to give a bit more context for *why* things are like that.

**Comment by oblitum** [comment]
If nanotee is correct, then solely having a `setup` implies a `opt = true`, while still being implicitly auto lazily loaded in the end. I wonder whether setting `opt = true` explicitly will make it actually "manual lazy".

**Reply by oblitum** [reply]
Ugh, indeed, this is exactly what happens. I think I know how it works now.
